tests/rptest/services/redpanda.py

```diff
@@ -1187,15 +1194,44 @@ def get_version(self, node):
         assert len(version_lines) == 1, version_lines
         return VERSION_LINE_RE.findall(version_lines[0])[0]
 
-    def stop_node(self, node, timeout=None):
-        pids = self.pids(node)
-
-        for pid in pids:
-            node.account.signal(pid, signal.SIGTERM, allow_fail=False)
+    def stop_node(self, node, timeout=None, enable_maintenance=False):
+        """
+        Stops the given node. If enable_maintenance is specified, puts the node
+        into maintenance mode and waits for the cluster to become healthy
+        following the shutdown.
+        """
+        def _has_drained(node):
+            """
+            Returns True if the node is finished draining.
+            """
+            try:
+                node_id = self.idx(node)
+                broker_resp = self._admin.get_broker(node_id)
+                maintenance_status = broker_resp["maintenance_status"]
+                return maintenance_status["draining"] and maintenance_status[
+                    "finished"]
+            except requests.exceptions.HTTPError:
+                return False
 
         if timeout is None:
             timeout = 30
 
+        if enable_maintenance:
+            wait_until(lambda: self.healthy(),
+                       timeout_sec=timeout,
+                       backoff_sec=1)
+            self._admin.maintenance_start(node)
+            wait_until(lambda: _has_drained(node),
+                       timeout_sec=timeout,
+                       backoff_sec=1)
+            wait_until(lambda: self.healthy(),
+                       timeout_sec=timeout,
+                       backoff_sec=1)
```

Comment:
Yeah I'm torn, I actually originally went the route of writing a `RollingRestarter` that operated on a `RedpandaService`, but ended up here because my requirements were simple and baking it into the service is an easy way to inform callers that this exists as an option to consider (it's easy to be tempted to write rolling restart from scratch using existing tools, but you probably wouldn't be tempted if you saw this argument in `restart_nodes()`!).

In general I worry that rolling restart is probably going to be the default way restarts are done moving forward, so having it be a first class test primitive seems worthwhile. Having an explicit `RollingRestarter` means folks need to be aware of it and opt in.

That said, it also means that we can extend it to be pretty sophisticated if we want, e.g. run in a background thread, run with rack awareness, etc. These seem like they'll be pretty important as we grow. I feel myself swinging back in this direction. I'll consider it..

Reply:
I'll keep the interface the same, but pull the bits into a separate class.